src/mixins/java/org/spongepowered/common/mixin/api/mcp/world/IWorldMixin_API.java

```diff
@@ -191,158 +218,179 @@ default WorldProperties getProperties() {
+                .createSpawnEntityEventCustom(Sponge.getCauseStackManager().getCurrentCause(), entitiesToSpawn);
+        if (Sponge.getEventManager().post(event)) {
+            return ImmutableList.of();
+        }
+        for (final Entity entity : event.getEntities()) {
+            EntityUtil.processEntitySpawn(entity, Optional::empty);
+        }
 
-    @Override
-    default boolean validateRawData(final int x, final int y, final int z, final DataView container) {
-        return false;
+        final ImmutableList.Builder<Entity> builder = ImmutableList.builder();
+        for (final Entity entity : event.getEntities()) {
+            builder.add(entity);
+        }
+        return builder.build();
     }
 
     @Override
-    default void setRawData(final int x, final int y, final int z, final DataView container) throws InvalidDataException {
+    default boolean spawnEntity(Entity entity) {
+        return IWorldWriterMixin_API.super.spawnEntity(entity); // TODO is capature needed here?
 
+        Preconditions.checkNotNull(entity, "The entity cannot be null!");
+        if (PhaseTracker.isEntitySpawnInvalid(entity)) {
+            return true;
+        }
+        final PhaseTracker phaseTracker = PhaseTracker.getInstance();
+        final IPhaseState<?> state = phaseTracker.getCurrentState();
+        if (!state.alreadyCapturingEntitySpawns()) {
+            try (final BasicPluginContext context = PluginPhase.State.CUSTOM_SPAWN.createPhaseContext(PhaseTracker.SERVER)) {
+                context.buildAndSwitch();
+                phaseTracker.spawnEntityWithCause((org.spongepowered.api.world.World<?>) this, entity);
+                return true;
+            }
+        }
+        return phaseTracker.spawnEntityWithCause((org.spongepowered.api.world.World<?>) this, entity);
     }
 
-    @Override
-    default int getSeaLevel() {
-        return 0;
-    }
+    // HeightAwareVolume
 
     @Override
-    default ScheduledUpdateList<BlockType> getScheduledBlockUpdates() {
-        return null;
+    default int getHeight(final HeightType type, final int x, final int z) {
+        return IWorldReaderMixin_API.super.getHeight(type, x, z);
     }
 
-    @Override
-    default ScheduledUpdateList<FluidType> getScheduledFluidUpdates() {
-        return null;
-    }
+    // MutableBlockVolume
 
     @Override
-    default boolean setBlock(final int x, final int y, final int z, final BlockState state, final BlockChangeFlag flag) {
-        return false;
+    default boolean removeBlock(int x, int y, int z) {
+        return IWorldGenerationReaderMixin_API.super.removeBlock(x, y, z);
     }
 
     @Override
-    default boolean spawnEntity(final Entity entity) {
-        return false;
-    }
+    default boolean setBlock(int x, int y, int z, org.spongepowered.api.block.BlockState blockState, BlockChangeFlag flag) {
+        return IWorldGenerationReaderMixin_API.super.setBlock(x, y, z, blockState, flag); // TODO is capture needed here?
```

Comment:
@gabizou api7 code was capturing stuff in this method. Is this needed here?
Otherwise `IWorldGenerationReaderMixin_API` already has the redirect to the `shadow$setBlockState`